### PR TITLE
Integrated C tensorflow API for dct

### PR DESCRIPTION
**Pull request by adanlopeza**
Implemented dct with tensorflow to improve performance.

### CODE REVIEW

File: xla/jfdctxla.c [hunk 1]
```diff
@@ -11,76 +11,220 @@
...
+ *
+ * AssertOk throws a runtime_error if Tensorflow communicates an
+ *   exceptional status.
+ *
+ */
+void AssertOk(const TF_Status* status);
+
+/**
+ * binary_graphdef_protobuf_filename: only binary protobuffers
+ *   seem to be supported via the tensorflow C api.
```
Comment (ghost) [inline xla/jfdctxla.c]: Confirm if this is the case and switch to "are supported"

File: xla/jfdctxla.c [hunk 2]
```diff
@@ -11,76 +11,220 @@
...
+ * input_node_name: the name of the node that should be feed with the
```
Comment (ghost) [inline xla/jfdctxla.c]: fed

File: xla/jfdctxla.c [hunk 3]
```diff
@@ -11,76 +11,220 @@
...
+    const char* input_node_name,
+    const char* output_node_name);
+
+/**
+ * Clean up all pointer-members using the dedicated tensorflor api functions
+ */
+void destroy(Inference* inf);
+
+/**
+ * Run the graph on some input data.
```
Comment (ghost) [inline xla/jfdctxla.c]: Which one of the two? Perhaps:

Run graph on input data and return input and output tensors.

File: xla/jfdctxla.c [hunk 4]
```diff
@@ -11,76 +11,220 @@
...
+ *
+ * Provide the input and output tensor.
+ */
+TF_Tensor* runGraph(const Inference* inf, TF_Tensor* input_tensor);
+
+/**
+ * Read a binary protobuf (.pb) buffer into a TF_Buffer object.
```
Comment (ghost) [inline xla/jfdctxla.c]: remove buffer

File: xla/jfdctxla.c [hunk 5]
```diff
@@ -11,76 +11,220 @@
...
+  }
+  struct stat stat;
+  if (fstat(fd, &stat) != 0) {
+    perror("failed to read file: ");
+    return NULL;
+  }
+  char* data = (char*)malloc(stat.st_size);
+  ssize_t nread = read(fd, data, stat.st_size);
+  if (nread < 0) {
+    perror("failed to read file: ");
```
Comment (ghost) [inline xla/jfdctxla.c]: Can you make the error message more accurate? (I.e. highlight difference between this and previous)

File: xla/jfdctxla.c [hunk 6]
```diff
@@ -11,76 +11,220 @@
...
-			}
-		}
-	}
+Inference* newInference(
+  const char* binary_graphdef_protobuffer_filename,
+  const char* input_node_name,
+  const char* output_node_name)
+{
+  Inference* obj = malloc(sizeof(Inference));
+  // init the 'trival' members
```
Comment (ghost) [inline xla/jfdctxla.c]: Start comment with capital letter, same below.

File: xla/jfdctxla.c [hunk 7]
```diff
@@ -11,76 +11,220 @@
...
+  TF_Status* status = TF_NewStatus();
+  obj->graph = TF_NewGraph();
+
+  // create a bunch of objects we need to init graph and session
```
Comment (ghost) [inline xla/jfdctxla.c]: Create objects needed to init graph and session: (Explain here briefly what they do)

File: xla/jfdctxla.c [hunk 8]
```diff
@@ -11,76 +11,220 @@
...
+  // prepare the constants for inference
+  // input
+  obj->input_op = TF_GraphOperationByName(obj->graph, input_node_name);
+  obj->input = newOutput(obj->input_op, 0);
+
+  // output
+  obj->output_op = TF_GraphOperationByName(obj->graph, output_node_name);
+  obj->output = newOutput(obj->output_op, 0);
+
+  // Clean Up all temporary objects
```
Comment (ghost) [inline xla/jfdctxla.c]: Clean up all temporary objects

File: xla/jfdctxla.c [hunk 9]
```diff
@@ -11,76 +11,220 @@
...
+  TF_SessionOptions* session_opts = TF_NewSessionOptions();
+
+  // import graph
+  TF_GraphImportGraphDef(obj->graph, graph_def, opts, status);
+  AssertOk(status);
+  // and create session
+  obj->session = TF_NewSession(obj->graph, session_opts, status);
+  AssertOk(status);
+
+  // prepare the constants for inference
```
Comment (ghost) [inline xla/jfdctxla.c]: Prepare? What specific preparation is done to the constants?

File: xla/jfdctxla.c [hunk 10]
```diff
@@ -11,76 +11,220 @@
...
+  TF_DeleteSessionOptions(session_opts);
+
+  TF_DeleteStatus(status);
+  return obj;
+}
+
+void destroy(Inference* inf)
+{
+  TF_Status* status = TF_NewStatus();
+  // Clean up all the members
```
Comment (ghost) [inline xla/jfdctxla.c]: Clean up all members

File: xla/jfdctxla.c [hunk 11]
```diff
@@ -11,76 +11,220 @@
...
+  TF_Output* input;
+
+  TF_Operation* output_op;
+  TF_Output* output;
+
+
+} Inference;
+
+/**
+ * Load a protobuf buffer from disk,
```
Comment (ghost) [inline xla/jfdctxla.c]: Why is the line so short?

File: xla/jfdctxla.c [hunk 12]
```diff
@@ -11,76 +11,220 @@
...
+
+TF_Output* newOutput(TF_Operation* oper, int index) {
+  TF_Output* obj = malloc(sizeof(TF_Output));
+  obj->oper = oper;
+  obj->index = index;
+  return obj;
+}
+
+/**
+ * Load a protobuf buffer from disk,
```
Comment (ghost) [inline xla/jfdctxla.c]: Short lines?

Comment (ghost) [reply xla/jfdctxla.c]: protobuf buffer is redundant. What information is contained in the protobuf?

File: xla/jfdctxla.c [hunk 13]
```diff
@@ -11,76 +11,220 @@
...
+void destroy(Inference* inf)
+{
+  TF_Status* status = TF_NewStatus();
+  // Clean up all the members
+  TF_CloseSession(inf->session, status);
+  TF_DeleteGraph(inf->graph);
+  TF_DeleteSession(inf->session, status);
+
+  TF_DeleteStatus(status);
+  // input_op & output_op are delete by deleting the graph
```
Comment (ghost) [inline xla/jfdctxla.c]: This comment seems out of place here. Also:

input_op & output_op are "delete", which deletes the graph.

Comment (ghost) [reply xla/jfdctxla.c]: Which graph, the whole graph? Why do you need to do it twice at input and output then?

File: xla/jfdctxla.c [hunk 14]
```diff
@@ -11,76 +11,220 @@
...
-		data[i] *= flow_coeff[i];
+  Inference* CNN = newInference("tfdct.pb", "x", "y");
+
+  // create an input tensor
+  int64_t dims[] = {8, 8};
+  TF_Tensor* in  = TF_AllocateTensor(TF_FLOAT, dims, 2, 8 * 8 * sizeof(float));
+  float* in_data = (float*)(TF_TensorData(in));
+  for(int i = 0; i < 8; i++)
+    for(int j = 0; j < 8; j++)
+      in_data[i * 8 + j] = data[i * 8 + j];
```
Comment (ghost) [inline xla/jfdctxla.c]: You are making a copy? Isn't there some simpler way?

File: xla/pbgenerator.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
+import os
+path = os.getcwd()
+
+# A trivial example for a start
+import tensorflow as tf
+
+tf.compat.v1.disable_eager_execution()
+sess=tf.compat.v1.InteractiveSession()
+
+dct_matrix = tf.compat.v1.constant(
```
Comment (ghost) [inline xla/pbgenerator.py]: How was this matrix generated? Please explain.

File: xla/pbgenerator.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+  [0.353553391, 0.353553391, 0.353553391, 0.353553391, 0.353553391, 0.353553391, 0.353553391, 0.353553391,
+   0.490392640, 0.415734806, 0.277785117, 0.097545161, -0.097545161, -0.277785117, -0.415734806, -0.490392640,
+   0.461939766, 0.191341716, -0.191341716, -0.461939766, -0.461939766, -0.191341716, 0.191341716, 0.461939766,
+   0.415734806, -0.097545161, -0.490392640, -0.277785117, 0.277785117, 0.490392640, 0.097545161, -0.415734806,
+   0.353553391, -0.353553391, -0.353553391, 0.353553391, 0.353553391, -0.353553391, -0.353553391, 0.353553391,
+   0.277785117, -0.490392640, 0.097545161, 0.415734806, -0.415734806, -0.097545161, 0.490392640, -0.277785117,
+   0.191341716, -0.461939766, 0.461939766, -0.191341716, -0.191341716, 0.461939766, -0.461939766, 0.191341716,
+   0.097545161, -0.277785117, 0.415734806, -0.490392640, 0.490392640, -0.415734806, 0.277785117, -0.097545161], shape=[8, 8], dtype=tf.float32)
+
+dct_transp = tf.compat.v1.constant([0.353553391, 0.490392640, 0.461939766, 0.415734806, 0.353553391, 0.277785117, 0.191341716, 0.097545161,
```
Comment (ghost) [inline xla/pbgenerator.py]: 80 character limit

File: xla/pbgenerator.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+      0.353553391, 0.415734806, 0.191341716, -0.097545161, -0.353553391, -0.490392640, -0.461939766, -0.277785117,
+      0.353553391, 0.277785117, -0.191341716, -0.490392640, -0.353553391, 0.097545161, 0.461939766, 0.415734806,
+      0.353553391, 0.097545161, -0.461939766, -0.277785117, 0.353553391, 0.415734806, -0.191341716, -0.490392640,
+      0.353553391, -0.097545161, -0.461939766, 0.277785117, 0.353553391, -0.415734806, -0.191341716, 0.490392640,
+      0.353553391, -0.277785117, -0.191341716, 0.490392640, -0.353553391, -0.097545161, 0.461939766, -0.415734806,
+      0.353553391, -0.415734806, 0.191341716, 0.097545161, -0.353553391, 0.490392640, -0.461939766, 0.277785117,
+      0.353553391, -0.490392640, 0.461939766, -0.415734806, 0.353553391, -0.277785117, 0.191341716, -0.097545161], dtype=tf.float32, shape=[8, 8])
+
+flow_coeff = tf.compat.v1.constant([8.000000,  11.096321, 10.452571, 9.406975,  8.000002,  6.285559,  4.329570,  2.207195,
```
Comment (ghost) [inline xla/pbgenerator.py]: Same as above, please explain origin of matrix.

File: xla/pbgenerator.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+                                      9.407003,  13.047894, 12.290842, 11.061483, 9.407002,  7.391037,  5.091034,  2.595389,
+                                      8.000001,  11.096319, 10.452504, 9.407007,  8.000000,  6.285560,  4.329569,  2.207195,
+                                      6.285562,  8.718323,  8.212480,  7.391035,  6.285561,  4.938532,  3.401721,  1.734182,
+                                      4.329566,  6.005280,  5.656854,  5.091043,  4.329567,  3.401721,  2.343146,  1.194525,
+                                      2.207195,  3.061468,  2.883839,  2.595387,  2.207199,  1.734182,  1.194525,  0.608964], dtype=tf.float32, shape=[8, 8])
+
+# This placeholder is where the inference in C++ starts. Therefore the name
+# set here will be used later as 'input_node_name'
+x   = tf.compat.v1.placeholder(dtype=tf.float32, shape=[8, 8], name="x")
+y   = tf.compat.v1.multiply(tf.compat.v1.matmul(tf.compat.v1.matmul(dct_matrix, x), dct_transp), flow_coeff, name="y")
```
Comment (ghost) [inline xla/pbgenerator.py]: Define intermediate value for readability.

dct_x_dct_transp = tf.compat.v1.matmul(tf.compat.v1.matmul(dct_matrix, x), dct_transp)

(Again, watch 80 character limit, seems to be exceed all over this file)

File: xla/pbgenerator.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+# 'y' is where the inference ends: output_node_name = 'y'
+
+# try the graph
+print("Sample matrix: ", sess.run(y, feed_dict={x:[
+    [  58.05524177 ,  16.89659618 ,  53.58510851 ,   6.9319865 ,  105.89038398,
+                                                   -17.46786134 , -63.75375256 ,  38.18825511],
+                                                [ 109.98899245 , -58.39964431 , -97.64942703 , -44.78269159 , -29.26760241,
```
Comment (ghost) [inline xla/pbgenerator.py]: Please align rows and keep 80 char lim

File: xla/pbgenerator.py [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+                                                    -53.77262291 ,  73.30913674 , 103.13819522],
+                                                [ -23.32278361 ,  12.06769763 , -87.41577477 , -50.91094088 ,  67.46144819,
+                                                     37.36131768 ,-124.30599307 ,   3.4378598 ],
+                                                [ -82.52597314 ,  93.78118258 ,  60.53115988 ,  46.87563445 , -47.95005221,
+                                                    -78.41534865 , 121.57929678 ,  92.96126827],
+                                                [ -97.330442 ,    67.19502772 , -53.66041295 , -84.04221383 ,  98.32921252,
+                                                    -32.54579311 ,-117.54512154 , -54.82065896]]}))
+
+# save the graph
+tf.compat.v1.train.write_graph(graph_or_graph_def = sess.graph_def,
```
Comment (ghost) [inline xla/pbgenerator.py:47]: Please align arguments.

File: xla/pbgenerator.py [hunk 21]
```diff
@@ -0,0 +1,69 @@
+import os
```
Comment (ghost) [inline xla/pbgenerator.py:3]: Missing high level description of what this file does.

File: xla/pbgenerator.py [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+flow_coeff = tf.compat.v1.constant([8.000000,  11.096321, 10.452571, 9.406975,  8.000002,  6.285559,  4.329570,  2.207195,
+                                      11.096320, 15.391046, 14.498035, 13.047892, 11.096315, 8.718322,  6.005284,  3.061468,
+                                      10.452501, 14.498041, 13.656854, 12.290866, 10.452503, 8.212481,  5.656853,  2.883838,
+                                      9.407003,  13.047894, 12.290842, 11.061483, 9.407002,  7.391037,  5.091034,  2.595389,
+                                      8.000001,  11.096319, 10.452504, 9.407007,  8.000000,  6.285560,  4.329569,  2.207195,
+                                      6.285562,  8.718323,  8.212480,  7.391035,  6.285561,  4.938532,  3.401721,  1.734182,
+                                      4.329566,  6.005280,  5.656854,  5.091043,  4.329567,  3.401721,  2.343146,  1.194525,
+                                      2.207195,  3.061468,  2.883839,  2.595387,  2.207199,  1.734182,  1.194525,  0.608964], dtype=tf.float32, shape=[8, 8])
+
+# This placeholder is where the inference in C++ starts. Therefore the name
```
Comment (ghost) [inline xla/pbgenerator.py]: C ++ or C?

**Review by ghost — CHANGES_REQUESTED** [review]
Files looks good overall. I only had minor comments on style. 

Missing unit tests though, need to add them before submission. It seems like your pbgenerator.py file could be changed into unit tests easily.

File: CMakeLists.txt [hunk 23]
```diff
@@ -560,15 +560,16 @@ endif()
 
 if(ENABLE_STATIC)
   add_library(jpeg-static STATIC ${JPEG_SOURCES} $<TARGET_OBJECTS:simd>
-    ${SIMD_OBJS} $<TARGET_OBJECTS:xla>)
+    ${SIMD_OBJS})
+  target_link_libraries(jpeg-static xla)
```
Comment (itaiz-google) [inline CMakeLists.txt:564]: I don't understand why you made these changes.

File: xla/jfdctxla.c [hunk 24]
```diff
@@ -11,76 +11,53 @@
...
 {
-	FAST_FLOAT tmp[DCTSIZE * DCTSIZE];
-	int i;
-
-	matrix_multiply(dct_matrix, data, tmp);
-	matrix_multiply(tmp, transposed_dct_matrix, data);
-	
-	for(i = 0; i < DCTSIZE * DCTSIZE; i++)
-		data[i] *= flow_coeff[i];
+  initialize_tf_session();
```
Comment (itaiz-google) [inline xla/jfdctxla.c]: Please do initialization outside of the DCT.

File: xla/CMakeLists.txt [hunk 25]
```diff
@@ -1,10 +1,19 @@
-set(XLA_SOURCES jfdctxla.c)
-add_library(xla OBJECT ${XLA_SOURCES})
+# External tensorflow library.
```
Comment (itaiz-google) [inline xla/CMakeLists.txt:1]: Is this only conditionally included?  It should be.

File: CMakeLists.txt (hunk 23 above)
Comment (adanlopeza) [reply CMakeLists.txt:564]: I'm rewriting changes I made in the last PR.

File: xla/CMakeLists.txt (hunk 25 above)
Comment (adanlopeza) [reply xla/CMakeLists.txt:1]: This file itself is conditionally included. Look at [this](https://github.com/googleinterns/libjpeg-turbo/blob/xla/CMakeLists.txt#L548) line.

**Review by itaiz-google — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**